### PR TITLE
Respect m.space.parent relations if they hold valid permissions

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/10935

---
Here's what your changelog entry will look like:

## ✨ Features
 * Respect m.space.parent relations if they hold valid permissions ([\#6746](https://github.com/matrix-org/matrix-react-sdk/pull/6746)). Fixes vector-im/element-web#10935.

Preview: https://613748c305bf375258f25cea--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by novocaine** [comment]
Can you confirm the user-facing effect of this change please? How would I test whether this is working correctly?

**Comment by t3chguy** [comment]
There is no UI to create "hidden" rooms at this time, but as per the MSC this is a possible use-case. You have a room `R` which has an `m.space.parent` event pointing at space `S` - the sender of this event must have permission in `S` to send `m.space.child` events, now clients will render `R` inside `S`.

**Comment by t3chguy** [comment]
There is no UI to create "hidden" rooms at this time, but as per the MSC this is a possible use-case. You have a room `R` which has an `m.space.parent` event pointing at space `S` - the sender of this event must have permission in `S` to send `m.space.child` events, now clients will render `R` inside `S`.

File: src/stores/SpaceStore.tsx [hunk 1]
```diff
@@ -366,16 +366,18 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
         const room = this.matrixClient?.getRoom(roomId);
         return room?.currentState.getStateEvents(EventType.SpaceParent)
-            .filter(ev => {
+            .map(ev => {
                 const content = ev.getContent();
-                if (!content?.via?.length) return false;
-                // TODO apply permissions check to verify that the parent mapping is valid
-                if (canonicalOnly && !content?.canonical) return false;
-                return true;
+                if (!Array.isArray(content?.via)) return;
```
Comment (novocaine) [inline src/stores/SpaceStore.tsx:374]: Suggestion: I find the `return;` here a bit hard to grok initially - this is a function that returns values, so I'd expect to see an explicit `return undefined`. Just at a glance it looks like the `returns` are strictly for bailing out rather than returning something that's going to be filtered on.

File: src/stores/SpaceStore.tsx [hunk 2]
```diff
@@ -366,16 +366,18 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
...
+                const parent = this.matrixClient.getRoom(ev.getStateKey());
+                if (canonicalOnly && !content?.canonical) return;
+                if (parent.currentState.maySendStateEvent(EventType.SpaceChild, userId)) {
```
Comment (novocaine) [inline src/stores/SpaceStore.tsx:377]: Comment: I find it pretty damn weird that the filter is "only show the linkage if you *could* have created this link now", but as noted in https://github.com/vector-im/element-web/issues/10935#issuecomment-898909964 that's what the spec says

File: src/stores/SpaceStore.tsx [hunk 3]
```diff
@@ -530,6 +532,14 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
             });
         }
 
+        const hiddenChildren = new EnhancedMap<string, Set<string>>();
```
Comment (novocaine) [inline src/stores/SpaceStore.tsx:535]: Suggestion: I found the name `hiddenChildren` a bit misleading as, if I understand correctly, it contains a { [ parent ]: visibleRoom } mapping for all rooms in `visibleRoom` - which is done for all visible rooms, not just for hidden ones .. maybe a name like `visibleRoomParents` or `parentsToVisibleRooms` would be closer to what it is?

**Comment by novocaine** [comment]
I think this works, I just found it a bit hard to parse so noted some suggestions for readability

File: src/stores/SpaceStore.tsx (hunk 2 above)
Comment (novocaine) [reply src/stores/SpaceStore.tsx:377]: Given its weird, it might make a lot of sense to leave a comment either quoting or linking to the part of the spec that specifies the behaviour

File: src/stores/SpaceStore.tsx (hunk 3 above)
Comment (t3chguy) [reply src/stores/SpaceStore.tsx:539]: Right but it is a map from `Parent -> Set<HiddenChild>`, so is the list of hidden children, per parent.

**Review by novocaine — APPROVED** [review]
:shipit:
